File: backend/app/schemas/series.py
"""剧集 (Series) API 的 Pydantic 数据模型。

包含创建、更新、响应三类载荷。字段名 / 类型与 ORM ``Series`` 保持一致；
``SeriesUpdate`` 把字段退化为可选，支持 PATCH 风格的局部更新。
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict


class SeriesBase(BaseModel):
    """Series 共享字段集合，供 Create 复用。

    Attributes:
        name: 剧集展示名。
        library_id: 关联的媒体库主键；可空表示未关联。
        emby_series_id: Emby 侧 Series ID；可空。
        staging_path: 整理缓存路径；须落在关联 library 的 staging_root 下。
        target_path: Emby 实际挂载路径；须落在关联 library 的 target_root 下。
        default_season: 默认季号，默认 1。
    """

    name: str
    library_id: int | None = None
    emby_series_id: str | None = None
    staging_path: str | None = None
    target_path: str | None = None
    default_season: int = 1


class SeriesCreate(SeriesBase):
    """``POST /api/v1/series`` 请求载荷。

    Attributes:
        enabled: 是否启用该 series；默认 ``True``。
    """

    enabled: bool = True


class SeriesUpdate(BaseModel):
    """``PUT /api/v1/series/{id}`` 请求载荷。

    所有字段可选；调用方按需传入需要修改的字段，未传字段保持原值。
    """

    name: str | None = None
    library_id: int | None = None
    emby_series_id: str | None = None
    staging_path: str | None = None
    target_path: str | None = None
    default_season: int | None = None
    enabled: bool | None = None


class SeriesResponse(BaseModel):
    """``GET / POST / PUT /api/v1/series`` 响应载荷。

    Attributes:
        id: 数据库主键。
        emby_series_id: Emby 侧 Series ID。
        name: 剧集展示名。
        library_id: 关联的媒体库主键。
        library_name: 关联媒体库名称（冗余字段，便于列表展示）。
        staging_path: 整理缓存路径。
        target_path: 目标路径。
        default_season: 默认季号。
        enabled: 是否启用。
        created_at: 创建时间。
        updated_at: 最后更新时间。
    """

    model_config = ConfigDict(from_attributes=True)

    id: int
    emby_series_id: str | None
    name: str
    library_id: int | None
    library_name: str | None
    staging_path: str | None
    target_path: str | None
    default_season: int
    enabled: bool
    created_at: datetime
    updated_at: datetime

File: frontend/src/stores/series.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'

import { api } from '@/api/client'

export interface Series {
  id: number
  emby_series_id: string | null
  name: string
  library_id: number | null
  library_name: string | null
  staging_path: string | null
  target_path: string | null
  default_season: number
  enabled: boolean
  created_at: string
  updated_at: string
}

export interface SeriesCreatePayload {
  name: string
  library_id?: number | null
  emby_series_id?: string | null
  staging_path?: string | null
  target_path?: string | null
  default_season: number
  enabled: boolean
}

export interface SeriesUpdatePayload {
  name?: string | null
  library_id?: number | null
  emby_series_id?: string | null
  staging_path?: string | null
  target_path?: string | null
  default_season?: number | null
  enabled?: boolean | null
}

export interface Season {
  Id?: string
  Name?: string
  IndexNumber?: number
  [key: string]: unknown
}

export interface Episode {
  Id?: string
  Name?: string
  IndexNumber?: number
  ParentIndexNumber?: number
  [key: string]: unknown
}

export interface EmbySeries {
  Id?: string
  Name?: string
  ProductionYear?: number
  [key: string]: unknown
}

export interface LatestEpisode {
  latest_episode: number
  next_episode: number
}

function getErrorMessage(error: unknown): string {
  return error instanceof Error ? error.message : '请求失败'
}

function withQuery(path: string, params: Record<string, string | number | undefined>) {
  const search = new URLSearchParams()
  for (const [key, value] of Object.entries(params)) {
    if (value !== undefined) search.set(key, String(value))
  }
  const query = search.toString()
  return query ? `${path}?${query}` : path
}

/**
 * 管理剧集配置与 Emby 剧集查询结果。
 *
 * @returns 剧集配置、Emby 季集数据状态与相关加载/保存动作。
 */
export const useSeriesStore = defineStore('series', () => {
  const series = ref<Series[]>([])
  const currentSeries = ref<Series | null>(null)
  const seasons = ref<Season[]>([])
  const episodes = ref<Episode[]>([])
  const latestEpisode = ref<LatestEpisode | null>(null)
  const embySearchResults = ref<EmbySeries[]>([])
  const loading = ref(false)
  const error = ref<string | null>(null)

  async function run<T>(action: () => Promise<T>): Promise<T | null> {
    loading.value = true
    error.value = null

    try {
      return await action()
    } catch (caught) {
      error.value = getErrorMessage(caught)
      return null
    } finally {
      loading.value = false
    }
  }

  async function loadSeries(libraryId?: number) {
    return run(async () => {
      const loadedSeries = await api.get<Series[]>(withQuery('/series', { library_id: libraryId }))
      series.value = loadedSeries
      return loadedSeries
    })
  }

  async function loadEmbySeries(q: string) {
    return run(async () => {
      const results = await api.get<EmbySeries[]>(withQuery('/emby/series/search', { keyword: q }))
      embySearchResults.value = results
      return results
    })
  }

  async function loadSeasons(embySeriesId: string) {
    return run(async () => {
      const loadedSeasons = await api.get<Season[]>(`/emby/series/${embySeriesId}/seasons`)
      seasons.value = loadedSeasons
      return loadedSeasons
    })
  }

  async function loadEpisodes(embySeriesId: string, season: number) {
    return run(async () => {
      const loadedEpisodes = await api.get<Episode[]>(`/emby/series/${embySeriesId}/seasons/${season}/episodes`)
      episodes.value = loadedEpisodes
      return loadedEpisodes
    })
  }

  async function loadLatestEpisode(embySeriesId: string, season: number) {
    return run(async () => {
      const episode = await api.get<LatestEpisode>(`/emby/series/${embySeriesId}/seasons/${season}/latest`)
      latestEpisode.value = episode
      return episode
    })
  }

  async function createSeriesConfig(payload: SeriesCreatePayload) {
    return run(async () => {
      const createdSeries = await api.post<Series>('/series', payload)
      series.value = [...series.value, createdSeries]
      currentSeries.value = createdSeries
      return createdSeries
    })
  }

  async function updateSeriesConfig(id: number, payload: SeriesUpdatePayload) {
    return run(async () => {
      const updatedSeries = await api.put<Series>(`/series/${id}`, payload)
      series.value = series.value.map((item) => (item.id === id ? updatedSeries : item))
      currentSeries.value = updatedSeries
      return updatedSeries
    })
  }

  return {
    series,
    currentSeries,
    seasons,
    episodes,
    latestEpisode,
    embySearchResults,
    loading,
    error,
    loadSeries,
    loadEmbySeries,
    loadSeasons,
    loadEpisodes,
    loadLatestEpisode,
    createSeriesConfig,
    updateSeriesConfig,
  }
})
